### PR TITLE
MBS-6978 copy artifacts even if build failed

### PR DESCRIPTION
**Pull request by int02h**

### CODE REVIEW

File: subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt [hunk 1]
```diff
@@ -51,13 +51,12 @@ open class VerifyArtifactsStep(
             entries.set(project.files(artifactsConfig.outputs.values.map { it.path }))
 
             project.gradle.buildFinished { buildResult ->
-                val isCopyScheduled = buildResult.gradle?.isTaskScheduled(this) ?: false
```
Comment (int02h) [inline subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt:54]: It seems that when a build is already finished there are no scheduled tasks, so this flag is always `false`.

File: subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt [hunk 2]
```diff
@@ -51,13 +51,12 @@ open class VerifyArtifactsStep(
...
-                // TODO: consider skipped state (--dry-run)
-                val copyFinished = this.didWork
-                if (isCopyScheduled && !copyFinished) {
-                    project.ciLogger.info("Artifacts copying was scheduled but didn't run. Start it now")
-                    if (!this.didWork) {
```
Comment (int02h) [inline subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt:59]: I am sure that double-check is not necessary here

File: subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt [hunk 3]
```diff
@@ -51,13 +51,12 @@ open class VerifyArtifactsStep(
...
-                        this.doAction()
+                if (buildResult.failure != null && buildResult.action == "Build") {
+                    if (!didWork) {
+                        // Copy artifacts that managed to be generated.
+                        // Do not verify them because it is last resort to save anything.
+                        project.ciLogger.info("Build failed. Trying to copy artifacts that managed to be generated.")
```
Comment (sergioboi) [inline subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt]: debug

File: subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt [hunk 4]
```diff
@@ -51,13 +51,12 @@ open class VerifyArtifactsStep(
...
             project.gradle.buildFinished { buildResult ->
-                val isCopyScheduled = buildResult.gradle?.isTaskScheduled(this) ?: false
-                // TODO: consider skipped state (--dry-run)
-                val copyFinished = this.didWork
-                if (isCopyScheduled && !copyFinished) {
-                    project.ciLogger.info("Artifacts copying was scheduled but didn't run. Start it now")
-                    if (!this.didWork) {
-                        this.doAction()
+                if (buildResult.failure != null && buildResult.action == "Build") {
+                    if (!didWork) {
```
Comment (sergioboi) [inline subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt]: What if we failed on the copy task?

File: subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt (hunk 3 above)
Comment (int02h) [reply subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt]: I've left the same log-level as it was before

File: subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt (hunk 4 above)
Comment (int02h) [reply subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt]: Build is already failed but you'll see a message from copy task too

File: subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt [hunk 5]
```diff
@@ -51,13 +51,12 @@ open class VerifyArtifactsStep(
...
 
             project.gradle.buildFinished { buildResult ->
-                val isCopyScheduled = buildResult.gradle?.isTaskScheduled(this) ?: false
-                // TODO: consider skipped state (--dry-run)
-                val copyFinished = this.didWork
-                if (isCopyScheduled && !copyFinished) {
-                    project.ciLogger.info("Artifacts copying was scheduled but didn't run. Start it now")
-                    if (!this.didWork) {
-                        this.doAction()
+                if (buildResult.failure != null && buildResult.action == "Build") {
```
Comment (eugene-krivobokov) [inline subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt]: It's good to hide this low-level knowledge about "action == "Build".

**Review by eugene-krivobokov — APPROVED** [review]

File: subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt (hunk 5 above)
Comment (int02h) [reply subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt]: Hide where?

Comment (eugene-krivobokov) [reply subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt]: For example, in an extension utility method.

We missed this one or two times already. `buildFinished` callback can be called multiple times. To distinguish a "real" build we've found only this workaround.

For me, the problem is similar to [a main process](https://github.com/int02h/primaree/blob/master/library/src/main/java/com/dpforge/primaree/ApplicationExtensions.kt) - workaround for lifecycle.

Comment (int02h) [reply subprojects/gradle/cd/src/main/kotlin/com/avito/ci/steps/VerifyArtifactsStep.kt]: Done here - 3da553495ac0194d7b1ae822e0d163d6d4af731a

File: subprojects/gradle/utils/src/main/kotlin/com/avito/utils/Gradles.kt [hunk 6]
```diff
@@ -0,0 +1,13 @@
+package com.avito.utils
+
+import org.gradle.api.invocation.Gradle
+
+fun Gradle.onBuildFailed(block: () -> Unit) {
+    buildFinished { buildResult ->
+        if (buildResult.failure != null && buildResult.action == ACTION_BUILD) {
```
Comment (sergioboi) [inline subprojects/gradle/utils/src/main/kotlin/com/avito/utils/Gradles.kt:7]: Tell me please, what actions may I see except ACTION_BUILD?

Comment (eugene-krivobokov) [reply subprojects/gradle/utils/src/main/kotlin/com/avito/utils/Gradles.kt:7]: `Configure`

https://docs.gradle.org/current/javadoc/org/gradle/BuildResult.html#getAction--

**Pull request merged**